inst/include/dust/dust.hpp

```diff
@@ -567,41 +556,63 @@ class Dust {
+                  const size_t n_particles, bool set_state) {
+    const size_t n = _particles.size() == 0 ? 0 : n_state_full();
+    Particle<T> p(pars, step);
+    if (n > 0 && p.size() != n) {
+      std::stringstream msg;
+      msg << "'pars' created inconsistent state size: " <<
+        "expected length " << n << " but created length " <<
+        p.size();
+      throw std::invalid_argument(msg.str());
+    }
+    if (_particles.size() == n_particles) {
+#ifdef _OPENMP
+      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+      for (size_t i = 0; i < n_particles; ++i) {
+        _particles[i].set_pars(p, set_state);
+      }
+    } else {
+      _particles.clear();
+      _particles.reserve(n_particles);
+      for (size_t i = 0; i < n_particles; ++i) {
+        _particles.push_back(p);
+      }
     }
-    initialise_index();
   }
 
   void initialise(const std::vector<pars_t>& pars, const size_t step,
-                  const size_t n_particles) {
-    // NOTE: we select the initialise function at runtime, but should
-    // always get it right. We might throw otherwise?
-    //
-    // We can throw here so need to make a new copy of particles.
-    std::vector<Particle<T>> particles;
-    particles.reserve(n_particles * _n_pars);
+                  const size_t n_particles, bool set_state) {
+        size_t n = _particles.size() == 0 ? 0 : n_state_full();
+    std::vector<Particle<T>> p;
     for (size_t i = 0; i < _n_pars; ++i) {
-      for (size_t j = 0; j < n_particles; ++j) {
-        particles.push_back(Particle<T>(pars[i], step));
+      p.push_back(Particle<T>(pars[i], step));
+      if (n > 0 && p.back().size() != n) {
+        std::stringstream msg;
+        msg << "'pars' created inconsistent state size: " <<
+          "expected length " << n << " but parameter set " << i + 1 <<
+          " created length " << p.back().size();
+        throw std::invalid_argument(msg.str());
       }
-      if (i > 0) {
-        const size_t n_old = particles.front().size();
-        const size_t n_new = particles.back().size();
-        if (n_old != n_new) {
-          std::stringstream msg;
-          msg << "Pars created different state sizes: pars " << i + 1 <<
-            " (of " << _n_pars << ") had length " << n_new <<
-            " but expected " << n_old;
-          throw std::invalid_argument(msg.str());
+      n = p.back().size(); // ensures all particles have same size
+    }
+    if (_particles.size() == _n_particles_total) {
+#ifdef _OPENMP
+      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+#endif
+      for (size_t i = 0; i < _n_particles_total; ++i) {
+        _particles[i].set_pars(p[i / n_particles], set_state);
+      }
+    } else {
+      _particles.clear();
+      _particles.reserve(n_particles * _n_pars);
+      for (size_t i = 0; i < _n_pars; ++i) {
+        for (size_t j = 0; j < n_particles; ++j) {
+          _particles.push_back(p[i]);
```

Comment:
Could also have openmp here if you s/reserve/resize/ and insert at `i * n_particles + j` (but I'm easy, not sure how much difference it would make anyway)

Reply:
I looked into this and it's hard if we can't default construct. there are some workarounds to make push_back work in parallel but they are all grim

Reply:
I previously tried one work-around using `#pragma critical` to merge lists made by each thread. Was no faster, and sometimes segfaulted in other environments

Reply:
(so I removed it)

Reply:
This makes me pleased I did not try! This never shows up as a big timesink normally